develop/items/custom-data-components.md

```diff
@@ -0,0 +1,267 @@
+
+```java
+stack.set(ModItems.CLICK_COUNT_COMPONENT, newValue);
+```
+
+This takes our component type and the value we want to set it to. In this case it will be our new click count.
+
+Let's add a few lines to our `use()` method to read the old click count, increase it by one and then set the updated click count.
+
+```java
+@Override
+public TypedActionResult<ItemStack> use(World world, PlayerEntity user, Hand hand) {
+    // spawn lightning
+    ...
+
+    // Update the click count
+    ItemStack stack = user.getStackInHand(hand);
+    int clickCount = stack.getOrDefault(ModItems.CLICK_COUNT_COMPONENT, 0);
+    stack.set(ModItems.CLICK_COUNT_COMPONENT, ++clickCount);
+
+    return TypedActionResult.success(stack);
+}
+```
+
+Now try starting the game and right clicking with the Lightning Stick. Count how many times you right click it. If you open up your inventory and look at the item again you should see that the usage number has gone up by the amount you clicked it.
+
+![Tooltip showing "Used 8 times"](/assets/develop/items/custom_component_3.png)
+
+## Removing the component value {#removing-component-value}
+
+You can also remove a component from your `ItemStack` if it is no longer needed. This is done by using the `remove()` method which takes your component type.
+
+```java
+stack.remove(ModItems.CLICK_COUNT_COMPONENT);
+```
+
+This method also returns the value of the component before being removed, so you can also use it like so:
+
+```java
+int oldCount = stack.remove(ModItems.CLICK_COUNT_COMPONENT);
+```
+
+## Advanced Data Components (maps) {#advanced-data-components}
+
+You may need to store multiple attributes in a single component. As a vanilla example, the `minecraft:food` component stores several values related to food, such as `nutrition`, `saturation`, `eat_seconds` and more. These components are referred to by the game as "map components".
+
+For map components, you must create a `record` class to store the data. This is the type we'll register in our component type and what we'll read and write when interacting with an `ItemStack`. Start by making a new record class in an appropriate package (you might want to make a new one called `component` for this).
+
+```java
+public record MyCustomComponent() {
+    
+}
+```
+
+Notice that there's a set of brackets after the class name. This is where we define the list of properties we want our component to have. Let's add a float and a boolean called `temperature` and `burnt` respectively.
+
+```java
+public record MyCustomComponent(float temperature, boolean burnt) {
+    
+}
+```
+
+Since we are defining a custom data structure there won't be a pre-existing `Codec` for our use case like with the [basic component](#basic-data-components). This means we're going to have to construct our own codec. Let's define one in our record class using a `RecordCodecBuilder` which we can reference once we register the component.
```

Comment:
I think for developers with less experience, it would be helpful to link the codec page here, since the `RecordCodecBuilder` syntax and codec in general can be confusing when seeing it for the first time.

Reply:
Good idea, I did link to the Codec page earlier but i'll highlight the specific section for Record codec builders